### PR TITLE
Ava GUI: User Profile Manager + Other Fixes

### PR DESCRIPTION
**Pull request by IsaacMarovitz**
**Changes:**
- Redesigned User Profile Manager
- Fixes janky `ListBox`
- Moved files in `Ryujinx.Ava.UI.Helper` to `Ryujinx.Ava.UI.Helpers`
- Remove unused views

Screenshot 2023-01-09 at 09 30 31
Screenshot 2023-01-09 at 09 30 34
Screenshot 2023-01-09 at 09 30 39
Screenshot 2023-01-09 at 09 30 45
Screenshot 2023-01-05 at 09 43 34
Screenshot_2023-01-09_at_18 28 11
Screenshot 2023-01-05 at 09 43 56
Screenshot 2023-01-05 at 09 43 51

Superceeds part of #3997

### CODE REVIEW

**Comment by gdkchan** [comment]
It would be nice if you could include screenshots showing the changes that have been made, it's useful for providing quick feedback about the design, and also for inclusion on the progress reports.

**Comment by AcK77** [comment]
I havent checked the code but that what I've noticed during my testing:

- On the User Profiles Manager, if I select a different profiles and then click "Manage Saves" sometime I can get:
```
00:00:14.462 |E| Application : Unhandled exception caught: System.ArgumentOutOfRangeException: Index was out of range. Must be non-negative and less than the size of the collection. (Parameter 'index')
   at System.Collections.Generic.List`1.get_Item(Int32)
   at System.Collections.ObjectModel.Collection`1.System.Collections.IList.get_Item(Int32)
   at Avalonia.Controls.Utils.IEnumerableUtils.ElementAt(IEnumerable items, Int32 index) in /_/src/Avalonia.Controls/Utils/IEnumerableUtils.cs:line 87
   at Avalonia.Controls.Presenters.ItemVirtualizerSimple.RecycleContainers() in /_/src/Avalonia.Controls/Presenters/ItemVirtualizerSimple.cs:line 383
   at Avalonia.Controls.Presenters.ItemVirtualizerSimple.ItemsChanged(IEnumerable items, NotifyCollectionChangedEventArgs e) in /_/src/Avalonia.Controls/Presenters/ItemVirtualizerSimple.cs:line 181
   at Avalonia.Controls.ItemsControl.Avalonia.Controls.Utils.ICollectionChangedListener.PostChanged(INotifyCollectionChanged sender, NotifyCollectionChangedEventArgs e) in /_/src/Avalonia.Controls/ItemsControl.cs:line 183
   at Avalonia.Controls.Utils.CollectionChangedEventManager.Entry.<Avalonia.Utilities.IWeakEventSubscriber<System.Collections.Specialized.NotifyCollectionChangedEventArgs>.OnEvent>g__Notify|6_0(INotifyCollectionChanged incc, NotifyCollectionChangedEventArgs args, List`1 listeners) in /_/src/Avalonia.Controls/Utils/CollectionChangedEventManager.cs:line 112
   at Avalonia.Controls.Utils.CollectionChangedEventManager.Entry.<>c__DisplayClass6_0.<Avalonia.Utilities.IWeakEventSubscriber<System.Collections.Specialized.NotifyCollectionChangedEventArgs>.OnEvent>b__1() in /_/src/Avalonia.Controls/Utils/CollectionChangedEventManager.cs:line 146
   at Avalonia.Win32.Win32Platform.WndProc(IntPtr hWnd, UInt32 msg, IntPtr wParam, IntPtr lParam) in /_/src/Windows/Avalonia.Win32/Win32Platform.cs:line 276
   at Avalonia.Win32.Interop.UnmanagedMethods.DispatchMessage(MSG& lpmsg)
   at Avalonia.Win32.Win32Platform.RunLoop(CancellationToken cancellationToken) in /_/src/Windows/Avalonia.Win32/Win32Platform.cs:line 210
   at Avalonia.Threading.Dispatcher.MainLoop(CancellationToken cancellationToken) in /_/src/Avalonia.Base/Threading/Dispatcher.cs:line 65
   at Avalonia.Controls.ApplicationLifetimes.ClassicDesktopStyleApplicationLifetime.Start(String[] args) in /_/src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs:line 120
   at Avalonia.ClassicDesktopStyleApplicationLifetimeExtensions.StartWithClassicDesktopLifetime[T](T builder, String[] args, ShutdownMode shutdownMode) in /_/src/Avalonia.Controls/ApplicationLifetimes/ClassicDesktopStyleApplicationLifetime.cs:line 209
   at Ryujinx.Ava.Program.Main(String[] args) in D:\a\Ryujinx\Ryujinx\Ryujinx.Ava\Program.cs:line 49
```
- It could be nice to get the buttons at the left bottom (Same for other views):
Before:
image
After:
image
Or at least, add a separator between the view content and the buttons maybe.

- Instead of using a green dot for selected user, maybe it could be nice to use the style as you use for the edit button, a grey circle background with a blue tick on it (still in the left corner).
image

- The selector in the Avatar window isn't really obvious, It could be nice to use the same as the one I suggested above:
image

- When you create a new profile, if you don't add a picture profile, it throw a popup, why not just goes directly in the picture profile selection instead?
image

- When you open the manager view, the selected profile is at the first place:
image
If you select another profile, the order stay the same:
image
If you edit the select one (or any other profile) and then goes back to the profiles list:
image
The order have changed. It should change while the profile is selected.

- There is another issue, We always have a "Close" button at the bottom, when you do something related to profiles, you have "Cancel" button to goes back. When you are in the save management views, there is a "<-" button on the top. Maybe things should be more consistent here.

- The Saves views should show some infos about the selected profiles too.
- I'm not sure sorting saves by size is something needed, but for what it cost, it's fine.

**Comment by AcK77** [comment]
Second pass:

## Profile Selector:

- When you open the Profile Manager, profiles are sorted alphabetically which is nice, but the selected profile should be at the first place:
image
- A border on the selected profile will made selected more obvious:
image
- Buttons have a strange white highlighting when you click on it (maybe not related to this PR):
Pbgzw8QZj3
- User Profiles list isn't aligned with the other objects, I know it's because the list have some padding and a lighter background, but the color is too light, so maybe add a little border (like the menus) could do things better, or removing the padding;
image

## Profile Editor:

- The UserId textbox is disabled instead of read-only, maybe in the futur it could be nice to be able to select the user id and be able to copy/paste it.
- The avatar should have a drop shadow or a blue border, or a lighter one: 
image
- If you changed the Nickname or the Avatar but clicking back without saving, it should warns you there is pending change and ask if you want to apply them or not.

## Firmware Avatar List:
- The pictures aren't sorted, so the list seems to be populated randomly, you can reproduce by trying to change the avatar for the firmware data multiple time, the list order is random.
- Buttons height aren't the same as other views.

## Create User:
- Same as the Profile Editor, there is no warning if you going back with pending changes.
- Nickname textbox doesn't have a placeholder.
- Avatar is just an empty dark grey box, which is strange, maybe it should have an empty user, for example:
image

## Save manager views:
- There are some controls with the ~~famous~~ lighter background, which made things a bit strange.
- The "Recover Lost Account" view should show something if there isn't found account to recover.

**Comment by gdkchan** [comment]
> When you open the Profile Manager, profiles are sorted alphabetically which is nice, but the selected profile should be at the first place

Why? What does it add?

File: Ryujinx.Ava/UI/Controls/NavigationDialogHost.axaml [hunk 1]
```diff
@@ -10,7 +10,8 @@
     x:Class="Ryujinx.Ava.UI.Controls.NavigationDialogHost"
     Focusable="True">
     <ui:Frame 
-        HorizontalAlignment="Stretch"
+        HorizontalAlignment="Stretch"
```
Comment (marysaka) [inline Ryujinx.Ava/UI/Controls/NavigationDialogHost.axaml:13]:
```suggestion
        HorizontalAlignment="Stretch"
```

File: Ryujinx.Ava/UI/Windows/MainWindow.axaml.cs [hunk 2]
```diff
@@ -573,18 +573,18 @@ public void UpdateGameMetadata(string titleId)
...
+            const float colorMultiple = 0.5f;
 
             Color progressFgColor = Color.FromRgb(dominantColor.R, dominantColor.G, dominantColor.B);
             Color progressBgColor = Color.FromRgb(
-                (byte)(dominantColor.R / ColorDivisor),
-                (byte)(dominantColor.G / ColorDivisor),
-                (byte)(dominantColor.B / ColorDivisor));
+                (byte)(dominantColor.R * colorMultiple),
+                (byte)(dominantColor.G * colorMultiple),
+                (byte)(dominantColor.B * colorMultiple));
```
Comment (marysaka) [inline Ryujinx.Ava/UI/Windows/MainWindow.axaml.cs:587]: Why was that changed?

File: Ryujinx.Ava/UI/Windows/MainWindow.axaml [hunk 3]
```diff
@@ -512,7 +512,8 @@
                             IsVisible="{Binding ShowLoadProgress}"
                             Text="{Binding LoadHeading}"
                             TextAlignment="Left"
-                            TextWrapping="Wrap" />
+                            TextWrapping="Wrap"
+                            MaxWidth="500" />
```
Comment (marysaka) [inline Ryujinx.Ava/UI/Windows/MainWindow.axaml:516]: why was that change?

File: Ryujinx.Ava/UI/Windows/MainWindow.axaml [hunk 4]
```diff
@@ -545,7 +545,8 @@
                             FontSize="18"
                             IsVisible="{Binding ShowLoadProgress}"
                             Text="{Binding CacheLoadStatus}"
-                            TextAlignment="Left" />
+                            TextAlignment="Left"
+                            MaxWidth="500" />
```
Comment (marysaka) [inline Ryujinx.Ava/UI/Windows/MainWindow.axaml:549]: why was that change?

**Review by marysaka — APPROVED** [review]

File: Ryujinx.Ava/UI/Windows/MainWindow.axaml (hunk 3 above)
Comment (IsaacMarovitz) [reply Ryujinx.Ava/UI/Windows/MainWindow.axaml]: To fix inconsistencies between different lines of text in the load view

File: Ryujinx.Ava/UI/Windows/MainWindow.axaml (hunk 4 above)
Comment (IsaacMarovitz) [reply Ryujinx.Ava/UI/Windows/MainWindow.axaml]: To fix inconsistencies between different lines of text in the load view

File: Ryujinx.Ava/UI/ViewModels/UserFirmwareAvatarSelectorViewModel.cs [hunk 5]
```diff
@@ -151,142 +84,71 @@ public byte[] SelectedImage
...
-
-                return streamJpg.ToArray();
+                image.BackgroundColor = new SolidColorBrush(BackgroundColor);
             }
         }
 
         public static void PreloadAvatars(ContentManager contentManager, VirtualFileSystem virtualFileSystem)
         {
-            try
+           if (_avatarStore.Count > 0)
```
Comment (gdkchan) [inline Ryujinx.Ava/UI/ViewModels/UserFirmwareAvatarSelectorViewModel.cs:107]: Indentation is off.

File: Ryujinx.Ava/UI/Views/User/UserFirmwareAvatarSelector.axaml.cs [hunk 6]
```diff
@@ -57,18 +62,25 @@ private void NavigatedTo(NavigationEventArgs arg)
...
         {
-            ViewModel.Dispose();
-
             _parent.GoBack();
         }
 
         private void ChooseButton_OnClick(object sender, RoutedEventArgs e)
         {
-            if (ViewModel.SelectedIndex > -1)
+            if (ViewModel.SelectedIndex > -1 && ViewModel.SelectedImage != null)
```
Comment (gdkchan) [inline Ryujinx.Ava/UI/Views/User/UserFirmwareAvatarSelector.axaml.cs:70]: Why are both checks needed? Can `SelectedImage` be not null but `SelectedIndex` be -1?

Comment (IsaacMarovitz) [reply Ryujinx.Ava/UI/Views/User/UserFirmwareAvatarSelector.axaml.cs:70]: I experienced a `NullReferenceException` at one point before I added this check

**Comment by AcK77** [comment]
Those 2 points have probably need to be discussed with other devs to know what they thinking about:

- When you open the Profile Manager, profiles are sorted alphabetically which is nice, but the selected profile should be at the first place:
image
- A border on the selected profile will made selected more obvious:
image

## Pending issues:

- Avatar are a bit small compared to GTK, but the loading is pretty faster and in the right order, the 6 items per row is pretty nice too:
image
- I'd invert the logic here, ask the user to saves pending change instead of discard it:
image
- "Recover Lost Accounts" button is still here for me even if the view is empty when I click on it:
image
image
- in the Save Manager, the list of save files should probably have the same little border as the list of profiles:
image
- Save items are smaller when they have an available save icon, which is not consistent:
image

**Comment by gdkchan** [comment]
> When you open the Profile Manager, profiles are sorted alphabetically which is nice, but the selected profile should be at the first place:

I don't think it's necessary for them to be the first profiles. It will make even less sense if we start allowing multiple profiles to be selected for multiple players. In a scenario where there are 2 players selecting their profiles with a controller, it would actually make selecting them more difficult because as soon as one players selects their profiles, the other ones would move potentially making the 2nd player select the wrong one. Of course, this is not a problem right now as we only allow one profile to be selected and don't support controller input on the UI, but they are things to consider for the future.

> A border on the selected profile will made selected more obvious:

I don't mind it, but I don't mind how it looks right now either. Right now it is already clear enough which profile is selected for me. But adding a border is fine for me too.

> I'd invert the logic here, ask the user to saves pending change instead of discard it:

I think it's fine like that, if the user clicked to exit, them they most likely want to exit rather than saving changes. If they wanted to save they would save clicked in save. Also it might not be possible to save if the profile is in a invalid state (like name is empty or no picture was chosen).

> "Recover Lost Accounts" button is still here for me even if the view is empty when I click on it:

It is useful for discoverability. Even if it is empty, users will know that this option exists. If it was hidden when there's nothing to recover, most likely wouldn't know that this option exists and would not look for it if they do eventually need to recover saves.

**Comment by AcK77** [comment]
> I don't think it's necessary for them to be the first profiles. It will make even less sense if we start allowing multiple profiles to be selected for multiple players. In a scenario where there are 2 players selecting their profiles with a controller, it would actually make selecting them more difficult because as soon as one players selects their profiles, the other ones would move potentially making the 2nd player select the wrong one. Of course, this is not a problem right now as we only allow one profile to be selected and don't support controller input on the UI, but they are things to consider for the future.
Good point. I remove it ;)

> I don't mind it, but I don't mind how it looks right now either. Right now it is already clear enough which profile is selected for me. But adding a border is fine for me too.
If we allow selecting multiple players in future with a controller, then the selected color (included a potentiel border) shoud be different than the actual blue one, so it could do things more obvious.

> I think it's fine like that, if the user clicked to exit, them they most likely want to exit rather than saving changes. If they wanted to save they would save clicked in save. Also it might not be possible to save if the profile is in a invalid state (like name is empty or no picture was chosen).
Since we have a "Save" button, that make sense.

> It is useful for discoverability. Even if it is empty, users will know that this option exists. If it was hidden when there's nothing to recover, they most likely wouldn't know that this option exists and would not look for it if they do eventually need to recover saves.
I suggested to add a text in the view to say the list is empty, @IsaacMarovitz have done https://github.com/Ryujinx/Ryujinx/pull/4166/commits/5a14b2b545c9de8f115744d4d86287015bc25f3f instead, that why I noticed the issue.

**Comment by IsaacMarovitz** [comment]
> the selected color (included a potentiel border) shoud be different than the actual blue one

With how this is set up right now selection is synonymous with the active profile. It is impossible to have a non active profile selected and vice versa. In the future where using multiple users at once will be required this system will have to be adjusted anyway. I think future proofing for that or controller based navigation is out of scope for this PR.

**Comment by IsaacMarovitz** [comment]
The Recover User Accounts button appearing when there is nothing to recover is a bug. But I think I will follow Ack's original suggestion instead

File: Ryujinx.Ava/UI/Views/User/UserFirmwareAvatarSelectorView.axaml [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+                        HorizontalAlignment="Center" />
+                </ItemsPanelTemplate>
+            </ListBox.ItemsPanel>
+            <ListBox.Styles>
+                <Style Selector="ListBoxItem">
+                    <Setter Property="Margin" Value="5" />
+                    <Setter Property="CornerRadius" Value="4" />
+                    <Setter Property="Width" Value="78" />
+                    <Setter Property="MaxWidth" Value="78" />
+                    <Setter Property="MinWidth" Value="78" />
```
Comment (AcK77) [inline Ryujinx.Ava/UI/Views/User/UserFirmwareAvatarSelectorView.axaml:56]:
```suggestion
                    <Setter Property="CornerRadius" Value="4" />
                    <Setter Property="Width" Value="85" />
                    <Setter Property="MaxWidth" Value="85" />
                    <Setter Property="MinWidth" Value="85" />
```

File: Ryujinx.Ava/UI/Views/User/UserFirmwareAvatarSelectorView.axaml [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+                    <Setter Property="MinHeight" Value="70" />
+                </Style>
+            </ListBox.Styles>
+            <ListBox.ItemTemplate>
+                <DataTemplate>
+                    <Panel
+                        Background="{Binding BackgroundColor}"
+                        Margin="5"
+                        Height="60"
+                        Width="60">
```
Comment (AcK77) [inline Ryujinx.Ava/UI/Views/User/UserFirmwareAvatarSelectorView.axaml:68]:
```suggestion
                        Margin="5">
```

File: Ryujinx.Ava/UI/Controls/NavigationDialogHost.axaml.cs [hunk 9]
```diff
@@ -78,14 +90,127 @@ public void Navigate(Type sourcePageType, object parameter)
...
+                    }
+                }
+            }
+
+            foreach(var account in lostAccounts)
+            {
+                ViewModel.LostProfiles.Add(new UserProfile(new HLE.HOS.Services.Account.Acc.UserProfile(account, "", null), this));
+            }
+
+            ViewModel.Profiles.Add(new AddModel());
```
Comment (AcK77) [inline Ryujinx.Ava/UI/Controls/NavigationDialogHost.axaml.cs:156]: Can't you use `BaseModel` in place of `AddModel` and then remove the `AddModel` empty class file ?

File: Ryujinx.Ava/UI/Controls/NavigationDialogHost.axaml.cs [hunk 10]
```diff
@@ -78,14 +90,127 @@ public void Navigate(Type sourcePageType, object parameter)
...
+
+            var profiles = AccountManager.GetAllUsers().OrderBy(x => x.Name);
+
+            foreach (var profile in profiles)
+            {
+                ViewModel.Profiles.Add(new UserProfile(profile, this));
+            }
+
+            var saveDataFilter = SaveDataFilter.Make(programId: default, saveType: SaveDataType.Account,
+                default, saveDataId: default, index: default);
```
Comment (AcK77) [inline Ryujinx.Ava/UI/Controls/NavigationDialogHost.axaml.cs:121]:
```suggestion
            var saveDataFilter = SaveDataFilter.Make(programId: default, saveType: SaveDataType.Account, default, saveDataId: default, index: default);
```

File: Ryujinx.Ava/UI/Controls/NavigationDialogHost.axaml.cs [hunk 11]
```diff
@@ -78,14 +90,127 @@ public void Navigate(Type sourcePageType, object parameter)
...
+
+                    return;
+                }
+
+                AccountManager.OpenUser(profile.UserId);
+            }
+
+            var result =
+                await ContentDialogHelper.CreateConfirmationDialog(LocaleManager.Instance[LocaleKeys.DialogUserProfileDeletionConfirmMessage], "",
+                    LocaleManager.Instance[LocaleKeys.InputDialogYes], LocaleManager.Instance[LocaleKeys.InputDialogNo], "");
```
Comment (AcK77) [inline Ryujinx.Ava/UI/Controls/NavigationDialogHost.axaml.cs:185]:
```suggestion
            var result = await ContentDialogHelper.CreateConfirmationDialog(
                LocaleManager.Instance[LocaleKeys.DialogUserProfileDeletionConfirmMessage],
                "",
                LocaleManager.Instance[LocaleKeys.InputDialogYes],
                LocaleManager.Instance[LocaleKeys.InputDialogNo],
                "");
```

File: Ryujinx.Ava/UI/Models/AddModel.cs [hunk 12]
```diff
@@ -0,0 +1,9 @@
+using Ryujinx.Ava.UI.ViewModels;
+
+namespace Ryujinx.Ava.UI.Models
+{
+    public class AddModel : BaseModel
+    {
+
```
Comment (AcK77) [inline Ryujinx.Ava/UI/Models/AddModel.cs:7]:
```suggestion
```

File: Ryujinx.Ava/UI/Models/UserProfile.cs [hunk 13]
```diff
@@ -43,27 +47,59 @@ public string Name
...
 
+        private void UpdateBackground()
+        {
+            Avalonia.Application.Current.Styles.TryGetResource("ControlFillColorSecondary", out object color);
+
+            if (color is not null)
+            {
+                BackgroundColor = _profile.AccountState == AccountState.Open
+                    ? new SolidColorBrush((Color)color)
+                    : Brushes.Transparent;
```
Comment (AcK77) [inline Ryujinx.Ava/UI/Models/UserProfile.cs:96]:
```suggestion
                BackgroundColor = _profile.AccountState == AccountState.Open ? new SolidColorBrush((Color)color) : Brushes.Transparent;
```

File: Ryujinx.Ava/UI/ViewModels/UserFirmwareAvatarSelectorViewModel.cs [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+        public static void PreloadAvatars(ContentManager contentManager, VirtualFileSystem virtualFileSystem)
+        {
+            if (_avatarStore.Count > 0)
+            {
+                return;
+            }
+
+            string contentPath =
+                contentManager.GetInstalledContentPath(0x010000000000080A, StorageId.BuiltInSystem,
+                    NcaContentType.Data);
```
Comment (AcK77) [inline Ryujinx.Ava/UI/ViewModels/UserFirmwareAvatarSelectorViewModel.cs:114]:
```suggestion
            string contentPath = contentManager.GetInstalledContentPath(0x010000000000080A, StorageId.BuiltInSystem, NcaContentType.Data);
```

File: Ryujinx.Ava/UI/ViewModels/UserFirmwareAvatarSelectorViewModel.cs [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+                using (IStorage ncaFileStream = new LocalStorage(avatarPath, FileAccess.Read, FileMode.Open))
+                {
+                    Nca nca = new(virtualFileSystem.KeySet, ncaFileStream);
+                    IFileSystem romfs = nca.OpenFileSystem(NcaSectionType.Data, IntegrityCheckLevel.ErrorOnInvalid);
+
+                    foreach (DirectoryEntryEx item in romfs.EnumerateEntries())
+                    {
+                        // TODO: Parse DatabaseInfo.bin and table.bin files for more accuracy.
+                        if (item.Type == DirectoryEntryType.File && item.FullPath.Contains("chara") &&
+                            item.FullPath.Contains("szs"))
```
Comment (AcK77) [inline Ryujinx.Ava/UI/ViewModels/UserFirmwareAvatarSelectorViewModel.cs:128]:
```suggestion
                        if (item.Type == DirectoryEntryType.File && item.FullPath.Contains("chara") && item.FullPath.Contains("szs"))
```

File: Ryujinx.Ava/UI/ViewModels/UserFirmwareAvatarSelectorViewModel.cs [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+                        {
+                            using var file = new UniqueRef<IFile>();
+
+                            romfs.OpenFile(ref file.Ref(), ("/" + item.FullPath).ToU8Span(), OpenMode.Read)
+                                .ThrowIfFailure();
```
Comment (AcK77) [inline Ryujinx.Ava/UI/ViewModels/UserFirmwareAvatarSelectorViewModel.cs:133]:
```suggestion
                            romfs.OpenFile(ref file.Ref(), ("/" + item.FullPath).ToU8Span(), OpenMode.Read).ThrowIfFailure();
```

File: Ryujinx.Ava/UI/Views/User/UserFirmwareAvatarSelectorView.axaml.cs [hunk 17]
```diff
@@ -53,22 +58,29 @@ private void NavigatedTo(NavigationEventArgs arg)
...
 
         private void ChooseButton_OnClick(object sender, RoutedEventArgs e)
         {
-            if (ViewModel.SelectedIndex > -1)
+            if (ViewModel.SelectedImage != null)
             {
-                _profile.Image = ViewModel.SelectedImage;
+                MemoryStream streamJpg = new();
+                SixLabors.ImageSharp.Image avatarImage =
+                    SixLabors.ImageSharp.Image.Load(ViewModel.SelectedImage, new PngDecoder());
```
Comment (AcK77) [inline Ryujinx.Ava/UI/Views/User/UserFirmwareAvatarSelectorView.axaml.cs:74]:
```suggestion
                SixLabors.ImageSharp.Image avatarImage = SixLabors.ImageSharp.Image.Load(ViewModel.SelectedImage, new PngDecoder());
```

File: Ryujinx.Ava/UI/Views/User/UserRecovererView.axaml.cs [hunk 18]
```diff
@@ -0,0 +1,52 @@
...
+            {
+                switch (arg.NavigationMode)
+                {
+                    case NavigationMode.New:
+                        var parent = (NavigationDialogHost)arg.Parameter;
+
+                        _parent = parent;
+
+                        ((ContentDialog)_parent.Parent).Title =
+                            $"{LocaleManager.Instance[LocaleKeys.UserProfileWindowTitle]} - {LocaleManager.Instance[LocaleKeys.UserProfilesRecoverHeading]}";
```
Comment (AcK77) [inline Ryujinx.Ava/UI/Views/User/UserRecovererView.axaml.cs:35]:
```suggestion
                        ((ContentDialog)_parent.Parent).Title = $"{LocaleManager.Instance[LocaleKeys.UserProfileWindowTitle]} - {LocaleManager.Instance[LocaleKeys.UserProfilesRecoverHeading]}";
```

File: Ryujinx.Ava/UI/Views/User/UserSaveManagerView.axaml.cs [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+        private void NavigatedTo(NavigationEventArgs arg)
+        {
+            if (Program.PreviewerDetached)
+            {
+                switch (arg.NavigationMode)
+                {
+                    case NavigationMode.New:
+                        var args =
+                            ((NavigationDialogHost parent, AccountManager accountManager, HorizonClient client, VirtualFileSystem
+                                virtualFileSystem))arg.Parameter;
```
Comment (AcK77) [inline Ryujinx.Ava/UI/Views/User/UserSaveManagerView.axaml.cs:50]:
```suggestion
                        var args = ((NavigationDialogHost parent, AccountManager accountManager, HorizonClient client, VirtualFileSystem virtualFileSystem))arg.Parameter;
```

File: Ryujinx.Ava/UI/Views/User/UserSaveManagerView.axaml.cs [hunk 20]
```diff
@@ -0,0 +1,129 @@
...
+                        _horizonClient = args.client;
+                        _virtualFileSystem = args.virtualFileSystem;
+
+                        _parent = args.parent;
+                        break;
+                }
+
+                DataContext = ViewModel = new UserSaveManagerViewModel(_accountManager);
+                ((ContentDialog)_parent.Parent).Title =
+                    $"{LocaleManager.Instance[LocaleKeys.UserProfileWindowTitle]} - {ViewModel.SaveManagerHeading}";
```
Comment (AcK77) [inline Ryujinx.Ava/UI/Views/User/UserSaveManagerView.axaml.cs:61]:
```suggestion
                ((ContentDialog)_parent.Parent).Title = $"{LocaleManager.Instance[LocaleKeys.UserProfileWindowTitle]} - {ViewModel.SaveManagerHeading}";
```

File: Ryujinx.Ava/UI/Views/User/UserSaveManagerView.axaml.cs [hunk 21]
```diff
@@ -0,0 +1,129 @@
...
+                Task.Run(LoadSaves);
+            }
+        }
+
+        public void LoadSaves()
+        {
+            ViewModel.Saves.Clear();
+            var saves = new ObservableCollection<SaveModel>();
+            var saveDataFilter = SaveDataFilter.Make(programId: default, saveType: SaveDataType.Account,
+                new UserId((ulong)_accountManager.LastOpenedUser.UserId.High, (ulong)_accountManager.LastOpenedUser.UserId.Low), saveDataId: default, index: default);
```
Comment (AcK77) [inline Ryujinx.Ava/UI/Views/User/UserSaveManagerView.axaml.cs:72]:
```suggestion
            var saveDataFilter = SaveDataFilter.Make(
                programId: default,
                saveType: SaveDataType.Account,
                new UserId((ulong)_accountManager.LastOpenedUser.UserId.High, (ulong)_accountManager.LastOpenedUser.UserId.Low),
                saveDataId: default,
                index: default);
```

File: Ryujinx.Ava/UI/Views/User/UserSelectorView.axaml.cs [hunk 22]
```diff
@@ -0,0 +1,127 @@
...
+using Ryujinx.Ava.UI.Controls;
+using Ryujinx.Ava.UI.ViewModels;
+using UserProfile = Ryujinx.Ava.UI.Models.UserProfile;
+
+namespace Ryujinx.Ava.UI.Views.User
+{
+    public partial class UserSelectorViews : UserControl
+    {
+        private NavigationDialogHost _parent;
+        public UserProfileViewModel ViewModel { get; set; }
```
Comment (AcK77) [inline Ryujinx.Ava/UI/Views/User/UserSelectorView.axaml.cs:16]:
```suggestion

        public UserProfileViewModel ViewModel { get; set; }
```

File: Ryujinx.Ava/Assets/Locales/en_US.json [hunk 23]
```diff
@@ -603,13 +607,15 @@
...
   "UserProfilesRecoverLostAccounts": "Recover Lost Accounts",
   "Recover": "Recover",
-  "UserProfilesRecoverHeading" : "Saves were found for the following accounts"
-}
-
+  "UserProfilesRecoverHeading" : "Saves were found for the following accounts",
+  "UserProfilesRecoverEmptyList": "No profiles to recover",
+  "UserEditorTitle" : "Edit User",
+  "UserEditorTitleCreate" : "Create User"
+}
```
Comment (AcK77) [inline Ryujinx.Ava/Assets/Locales/en_US.json:621]:
```suggestion
}

```

**Comment by AcK77** [comment]
Everything LGTM, except one little issue during my test, if you delete a save data, the list isn't refreshed automatically, you have to go back and list them again...

**Review by AcK77 — APPROVED** [review]
